Bare metal: APIs documentation update (#1)
rajkan01/mbed-os-5-docs: docs/api/api.md

<!--these are now just the APIs - do I want to add the tutorials?-->

This document lists Mbed OS components supported APIs. In general, it lists both common APIs (RTOS and the bare metal profile supported), and RTOS only supported APIs, but occasionally, It has either one.

## RTOS and event handling

The Mbed OS RTOS capabilities include managing objects such as threads, synchronization objects and timers. It also provides interfaces for attaching an application-specific idle hook function, reads the OS tick count and implements functionality to report RTOS errors.

Most of the Mbed OS APIs are also supported with the Mbed OS bare metal profile.

<table>
<tbody>
<tr>
<td><a href="semaphore.html">Semaphore</a></td>
<td><a href="userallocatedevent.html">UserAllocatedEvent</a></td>
<td><a href="eventflags.html">EventFlags</a></td>
<td><a href="kernel-interface-functions.html">Kernel interface functions</a></td>
</tr>
<tr>
<td><a href="thisthread.html">ThisThread</a></td>
<td><a href="event.html">Event</a></td>
<td><a href="eventqueue.html">EventQueue</a></td>
</tr>
</tbody>
</table>

### RTOS only supported APIs
<table>
<tbody>
<tr>
<td><a href="rtos-apis.html">Thread</a></td>

> <td><a href="mutex.html">Mutex</a></td>

Reviewer: Mutex is supported in bare mteal; it is a dummy implementation.

# Full API list
<!--these are now just the APIs - do I want to add the tutorials?-->

> This document lists Mbed OS components supported APIs. In general, it lists both common APIs (RTOS and the bare metal profile supported), and RTOS only supported APIs, but occasionally, It has either one.

Reviewer: ```suggestion
Mbed OS supports two profiles: the full profile, when an RTOS is present, and the bare metal profile which is a subset of the full profile.  In the list below, APIs which are supported in bare metal are noted with an asterik. A double asterisk indicates that the API is partially supported in bare metal.
```

## RTOS and event handling

> The Mbed OS RTOS capabilities include managing objects such as threads, synchronization objects and timers. It also provides interfaces for attaching an application-specific idle hook function, reads the OS tick count and implements functionality to report RTOS errors.

Reviewer: ```suggestion
The Mbed OS RTOS capabilities include managing objects such as threads and synchronization objects. It also provides interfaces for attaching an application-specific idle hook function, reads the OS tick count and implements functionality to report RTOS errors.
```

> Most of the Mbed OS APIs are also supported with the Mbed OS bare metal profile.

Reviewer: This sentence is redundant with the first one.

<table>
<tbody>
<tr>

> <td><a href="semaphore.html">Semaphore</a></td>

Reviewer: ```suggestion
<td><a href="semaphore.html">Semaphore<span>*</span></a></td>
```

> <td><a href="userallocatedevent.html">UserAllocatedEvent</a></td>

Reviewer: ```suggestion
<td><a href="userallocatedevent.html">UserAllocatedEvent<span>*</span></a></td>
```

> <td><a href="eventflags.html">EventFlags</a></td>

Reviewer: ```suggestion
<td><a href="eventflags.html">EventFlags<span>*</span></a></td>
```

> <td><a href="kernel-interface-functions.html">Kernel interface functions<br><i>(Only get_ms_count API)<i></a></td>

Reviewer: ```suggestion
<td><a href="kernel-interface-functions.html">Kernel<span>**</span></a></td>
```

</tr>
<tr>

> <td><a href="thisthread.html">ThisThread</a></td>

Reviewer: ```suggestion
<td><a href="thisthread.html">ThisThread<span>*</span></a></td>
```

> <td><a href="event.html">Event</a></td>

Reviewer: ```suggestion
<td><a href="event.html">Event<span>*</span></a></td>
```

> <td><a href="eventqueue.html">EventQueue</a></td>

Reviewer: ```suggestion
<td><a href="eventqueue.html">EventQueue<span>*</span></a></td>
```

> <td><a href="mutex.html">Mutex</a></td>

Reviewer: ```suggestion
<td><a href="mutex.html">Mutex<span>*</span></a></td>
```

</tr>

> </tbody>

Reviewer: ```suggestion
```

> </table>

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

> ### RTOS only supported APIs

Reviewer: ```suggestion
```

> <table>

Reviewer: ```suggestion
```

> <tbody>

Reviewer: ```suggestion
```

<tr>

> <td><a href="kernel-interface-functions.html">Kernel interface functions</a></td>

Reviewer: ```suggestion
```

<td><a href="rtos-apis.html">Thread</a></td>
<td><a href="queue.html">Queue</a></td>

> </tr>

Reviewer: ```suggestion
```

> <tr>

Reviewer: ```suggestion
```

<td><a href="mail.html">Mail</a></td>

> <td><a href="conditionvariable.html">Conditionvariable</a></td>

Reviewer: ```suggestion
<td><a href="conditionvariable.html">ConditionVariable</a></td>
```

The Mbed OS RTOS capabilities include managing objects such as threads, synchronization objects and timers. It also provides interfaces for attaching an application-specific idle hook function, reads the OS tick count and implements functionality to report RTOS errors.

> Most of the RTOS APIs are also compatible with Mbed OS bare metal profile. The Mbed OS bare metal profile is a compact profile of Mbed OS without an RTOS.

Reviewer: I'd just say "Most of the Mbed OS APIs are also supported with the Mbed OS bare metal profile."

> ### RTOS and the bare metal profile supported APIs

Reviewer: I don't think we should emphasize "RTOS and the bare metal profile supported APIs" - it is too messy. 
Instead I woud highlight the (few) RTOS only API.
